src/relay/charpane.ash

```diff
@@ -4532,112 +4532,112 @@ buffer addGroup(string[int] rowHTMLs, string className) {
-						default:
-							bricks[pLevel,i[pLevel]] = s;
-					}
-					break;
+
+
+buffer addBricks(string layout) {
+	buffer addBricksHelper (string[int] tokens) {
+	//parses a chit.__.layout string and adds bricks in appropriate rows and columns,
+	//with user-specified widths for the various bricks and groups of bricks
+	//uses multi-dimensional arrays for tracking arrays at different parenthesis depth.
+		buffer result; 
+		string[int,int] bricks;
+		string[int,int] rowHTML;
+		string[int,int] styleInfo;
+		string brick;
+		boolean read_style = false;
+		int pLevel = 0;      //tracks parenthesis depth
+		int[int] i = {0:0};  //tracks bricks within a row
+		int[int] j = {0:0};  //tracks rows withing a group
+		foreach n,s in tokens {
+			if(read_style && bricks[pLevel] contains i[pLevel]){
+				styleInfo[pLevel,i[pLevel]] = s;
+				read_style = false;
+			} else {
+				switch(s) {
+					case "": break;
+					case ":":
+					//next token tells us how wide this last element should be
+						read_style = true;
+						break;
+					case "|":
+					//next element will be horizontally adjacent to this one
+						i[pLevel]++;
+						break;
+					case ",":
+					//next element will start a new row below this one
+						rowHTML[pLevel,j[pLevel]] = addRow(bricks[pLevel], styleInfo[pLevel]);
+						i[pLevel]=0;
+						j[pLevel]++;
+						bricks[pLevel]={};
+						styleInfo[pLevel]={};
+						break;
+					case "{":
+					case "(":
+					//parenthesis depth increased. Pause current group parsing and start a new one.
+						pLevel++;
+						i[pLevel] = 0;
+						j[pLevel] = 0;
+						bricks[pLevel] = {};
+						styleInfo[pLevel] = {};
+						break;
+					case ")":
+					//vertical group complete. Add it to the list of bricks in the previous parenthesis depth and continue there.
+						rowHTML[pLevel,j[pLevel]] = addRow(bricks[pLevel], styleInfo[pLevel]);
+						bricks[pLevel-1,i[pLevel-1]] = addGroup(rowHTML[pLevel],"brick_column");
+						i[pLevel] = 0;
+						j[pLevel] = 0;
+						bricks[pLevel] = {};
+						styleInfo[pLevel] = {};
+						pLevel--;
+						break;
+					case "}":
+					//horizontal row complete. Add it to the list of bricks in the previous parenthesis depth and continue there.
+						rowHTML[pLevel,j[pLevel]] = addRow(bricks[pLevel], styleInfo[pLevel]);
+						bricks[pLevel-1,i[pLevel-1]] = addGroup(rowHTML[pLevel],"brick_row");
+						i[pLevel] = 0;
+						j[pLevel] = 0;
+						bricks[pLevel] = {};
+						styleInfo[pLevel] = {};
+						pLevel--;
+						break;
+					case "toolbar":
+					case "header":
+					case "footer":
+						break;	//Special Bricks that are inserted manually in the correct places
+					default:
+						if(chitBricks contains s && chitBricks[s] != "") {
+							bricks[pLevel,i[pLevel]] = chitBricks[s];
+						}
+						break;
+				}
 			}
 		}
+		if(pLevel != 0)
+			append(result, "<span>Malformed chit layout (mismatched parentheses): " + layout + "<span>");
```

Comment:
I was thinking more along the line of calling `error` than embedding the error straight in to the resulting charpane, but... I kind of like this more, so that's cool.

Second `<span>` should be `</span>` though.

Reply:
Yeah, I wrote this so I could see it working before I figured out how to throw an error message, and then I kinda liked it. It shows people which room the error is in, and if you're in a pre-rollover panic, you can ignore it for now and fix it later. I think it should probably get a <div class="error_message"> to wrap it, and then add some basic styling to the css file. I'll add that, and fix the </span> error.

Reply:
Looks nicer, but seems to no longer actually display the string value for some reason.

Reply:
Huh. I styled the string using "font-family: monospace" to give the a code-y look. Did it overflow to the next line, and then being cut off? I had it working on my machine.